manpages/en/fsck.fat.8
.\" fsck.fat.8 - manpage for fsck.fat
.\"
.\" Copyright (C) 2006-2013 Daniel Baumann <daniel@debian.org>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" The complete text of the GNU General Public License
.\" can be found in /usr/share/common-licenses/GPL-2 file.
.\"
.\"
.TH FSCK.FAT 8 2013\-07\-16 3.0.21 "dosfstools"

.SH NAME
\fBfsck.fat\fR \- check and repair MS\-DOS filesystems

.SH SYNOPSIS
\fBfsck.fat\fR|\fBfsck.msdos\fR|\fBfsck.vfat\fR [\-aAflnprtvVwy] [\-d \fIPATH\fR \-d\ \fI...\fR] [\-u\ \fIPATH\fR \-u \fI...\fR] \fIDEVICE\fR

.SH DESCRIPTION
\fBfsck.fat\fR verifies the consistency of MS\-DOS filesystems and optionally tries to repair them.
.PP
The following filesystem problems can be corrected (in this order):
.IP "*" 4
FAT contains invalid cluster numbers. Cluster is changed to EOF.
.IP "*" 4
File's cluster chain contains a loop. The loop is broken.
.IP "*" 4
Bad clusters (read errors). The clusters are marked bad and they are removed from files owning them. This check is optional.
.IP "*" 4
Directories with a large number of bad entries (probably corrupt). The directory can be deleted.
.IP "*" 4
Files . and .. are non\-directories. They can be deleted or renamed.
.IP "*" 4
Directories . and .. in root directory. They are deleted.
.IP "*" 4
Bad filenames. They can be renamed.
.IP "*" 4
Duplicate directory entries. They can be deleted or renamed.
.IP "*" 4
Directories with non\-zero size field. Size is set to zero.
.IP "*" 4
Directory . does not point to parent directory. The start pointer is adjusted.
.IP "*" 4
Directory .. does not point to parent of parent directory. The start pointer is adjusted.
.IP "*" 4
Start cluster number of a file is invalid. The file is truncated.
.IP "*" 4
File contains bad or free clusters. The file is truncated.
.IP "*" 4
File's cluster chain is longer than indicated by the size fields. The file is truncated.
.IP "*" 4
Two or more files share the same cluster(s). All but one of the files are truncated. If the file being truncated is a directory file that has already been read, the filesystem check is restarted after truncation.
.IP "*" 4
File's cluster chain is shorter than indicated by the size fields. The file is truncated.
.IP "*" 4
Clusters are marked as used but are not owned by a file. They are marked as free.
.PP
Additionally, the following problems are detected, but not repaired:
.IP "*" 4
Invalid parameters in boot sector.
.IP "*" 4
Absence of . and .. entries in non\-root directories
.PP
When \fBfsck.fat\fR checks a filesystem, it accumulates all changes in memory and performs them only after all checks are complete. This can be disabled with the \fB\-w\fR option.

.SH OPTIONS
.IP "\fB\-a\fR" 4
Automatically repair the filesystem. No user intervention is necessary. Whenever there is more than one method to solve a problem, the least destructive approach is used.
.IP "\fB\-A\fR" 4
Use Atari variation of the MS\-DOS filesystem. This is default if \fBfsck.fat\fR is run on an Atari, then this option turns off Atari format. There are some minor differences in Atari format: Some boot sector fields are interpreted slightly different, and the special FAT entries for end\-of\-file and bad cluster can be different. Under MS\-DOS 0xfff8 is used for EOF and Atari employs 0xffff by default, but both systems recognize all values from 0xfff8...0xffff as end\-of\-file. MS\-DOS uses only 0xfff7 for bad clusters, where on Atari values 0xfff0...0xfff7 are for this purpose (but the standard value is still 0xfff7).
.IP "\fB-b\fR" 4
Make read-only boot sector check.
.IP "\fB\-d\fR" 4
Delete the specified file. If more that one file with that name exists, the first one is deleted.
.IP "\fB\-f\fR" 4
Salvage unused cluster chains to files. By default, unused clusters are added to the free disk space except in auto mode (\fB\-a\fR).
.IP "\fB\-l\fR" 4
List path names of files being processed.
.IP "\fB\-n\fR" 4
No\-operation mode: non\-interactively check for errors, but don't write
anything to the filesystem.
.IP "\fB\-p\fR" 4
Same as (\fB\-a\fR), for compatibility with other *fsck.
.IP "\fB\-r\fR" 4
Interactively repair the filesystem. The user is asked for advice whenever
there is more than one approach to fix an inconsistency.
.IP "\fB\-t\fR" 4
Mark unreadable clusters as bad.
.IP "\fB\-u\fR" 4
Try to undelete the specified file. \fBfsck.fat\fR tries to allocate a chain of contiguous unallocated clusters beginning with the start cluster of the undeleted file.
.IP "\fB\-v\fR" 4
Verbose mode. Generates slightly more output.
.IP "\fB\-V\fR" 4
Perform a verification pass. The filesystem check is repeated after the first run. The second pass should never report any fixable errors. It may take considerably longer than the first pass, because the first pass may have generated long list of modifications that have to be scanned for each disk read.
.IP "\fB\-w\fR" 4
Write changes to disk immediately.
.IP "\fB\-y\fR" 4
Same as \fB\-a\fR (automatically repair filesystem) for compatibility with other fsck tools.
.PP
\fBNote:\fR If \fB\-a\fR and \fB\-r\fR are absent, the filesystem is only checked, but not repaired.

.SH "EXIT STATUS"
.IP "0" 4
No recoverable errors have been detected.
.IP "1" 4
Recoverable errors have been detected or \fBfsck.fat\fR has discovered an internal inconsistency.
.IP "2" 4
Usage error. \fBfsck.fat\fR did not access the filesystem.

.SH FILES
.IP "fsck0000.rec, fsck0001.rec, ..." 4
When recovering from a corrupted filesystem, \fBfsck.fat\fR dumps recovered data into files named 'fsckNNNN.rec' in the top level directory of the filesystem.

.SH BUGS
Does not create . and .. files where necessary. Does not remove entirely empty directories. Should give more diagnostic messages. Undeleting files should use a more sophisticated algorithm.

.SH SEE ALSO
\fBfatlabel\fR(8)
.br
\fBmkfs.fat\fR(8)

.SH HOMEPAGE
More information about \fBfsck.fat\fR and \fBdosfstools\fR can be found at <\fIhttp://daniel\-baumann.ch/software/dosfstools/\fR>.

.SH AUTHORS
\fBdosfstools\fR were written by Werner Almesberger <\fIwerner.almesberger@lrc.di.epfl.ch\fR>, Roman Hodek <\fIRoman.Hodek@informatik.uni-erlangen.de\fR>, and others. The current maintainer is Daniel Baumann <\fImail@daniel-baumann.ch\fR>.


src/fsck.fat.c
/* fsck.fat.c - User interface

   Copyright (C) 1993 Werner Almesberger <werner.almesberger@lrc.di.epfl.ch>
   Copyright (C) 1998 Roman Hodek <Roman.Hodek@informatik.uni-erlangen.de>
   Copyright (C) 2008-2013 Daniel Baumann <mail@daniel-baumann.ch>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   The complete text of the GNU General Public License
   can be found in /usr/share/common-licenses/GPL-2 file.
*/

/* FAT32, VFAT, Atari format support, and various fixes additions May 1998
 * by Roman Hodek <Roman.Hodek@informatik.uni-erlangen.de> */

#include "version.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>

#include "common.h"
#include "fsck.fat.h"
#include "io.h"
#include "boot.h"
#include "fat.h"
#include "file.h"
#include "check.h"
#include "charconv.h"

int interactive = 0, rw = 0, list = 0, test = 0, verbose = 0, write_immed = 0;
int atari_format = 0, boot_only = 0;
unsigned n_files = 0;
void *mem_queue = NULL;

static void usage(char *name)
{
    fprintf(stderr, "usage: %s [-aAbflrtvVwy] [-d path -d ...] "
	    "[-u path -u ...]\n%15sdevice\n", name, "");
    fprintf(stderr, "  -a       automatically repair the filesystem\n");
    fprintf(stderr, "  -A       toggle Atari filesystem format\n");
    fprintf(stderr, "  -b       make read-only boot sector check\n");
    fprintf(stderr,
	    "  -c N     use DOS codepage N to decode short file names (default: %d)\n",
	    DEFAULT_DOS_CODEPAGE);
    fprintf(stderr, "  -d path  drop that file\n");
    fprintf(stderr, "  -f       salvage unused chains to files\n");
    fprintf(stderr, "  -l       list path names\n");
    fprintf(stderr,
	    "  -n       no-op, check non-interactively without changing\n");
    fprintf(stderr, "  -p       same as -a, for compat with other *fsck\n");
    fprintf(stderr, "  -r       interactively repair the filesystem\n");
    fprintf(stderr, "  -t       test for bad clusters\n");
    fprintf(stderr, "  -u path  try to undelete that (non-directory) file\n");
    fprintf(stderr, "  -v       verbose mode\n");
    fprintf(stderr, "  -V       perform a verification pass\n");
    fprintf(stderr, "  -w       write changes to disk immediately\n");
    fprintf(stderr, "  -y       same as -a, for compat with other *fsck\n");
    exit(2);
}

/*
 * ++roman: On m68k, check if this is an Atari; if yes, turn on Atari variant
 * of MS-DOS filesystem by default.
 */
static void check_atari(void)
{
#ifdef __mc68000__
    FILE *f;
    char line[128], *p;

    if (!(f = fopen("/proc/hardware", "r"))) {
	perror("/proc/hardware");
	return;
    }

    while (fgets(line, sizeof(line), f)) {
	if (strncmp(line, "Model:", 6) == 0) {
	    p = line + 6;
	    p += strspn(p, " \t");
	    if (strncmp(p, "Atari ", 6) == 0)
		atari_format = 1;
	    break;
	}
    }
    fclose(f);
#endif
}

int main(int argc, char **argv)
{
    DOS_FS fs;
    int salvage_files, verify, c;
    unsigned n_files_check = 0, n_files_verify = 0;
    unsigned long free_clusters = 0;

    memset(&fs, 0, sizeof(fs));
    rw = salvage_files = verify = 0;
    interactive = 1;
    check_atari();

    while ((c = getopt(argc, argv, "Aac:d:bflnprtu:vVwy")) != EOF)
	switch (c) {
	case 'A':		/* toggle Atari format */
	    atari_format = !atari_format;
	    break;
	case 'a':
	case 'p':
	case 'y':
	    rw = 1;
	    interactive = 0;
	    salvage_files = 1;
	    break;
	case 'b':
	    rw = 0;
	    interactive = 0;
	    boot_only = 1;
	    break;
	case 'c':
	    set_dos_codepage(atoi(optarg));
	    break;
	case 'd':
	    file_add(optarg, fdt_drop);
	    break;
	case 'f':
	    salvage_files = 1;
	    break;
	case 'l':
	    list = 1;
	    break;
	case 'n':
	    rw = 0;
	    interactive = 0;
	    break;
	case 'r':
	    rw = 1;
	    interactive = 1;
	    break;
	case 't':
	    test = 1;
	    break;
	case 'u':
	    file_add(optarg, fdt_undelete);
	    break;
	case 'v':
	    verbose = 1;
	    printf("fsck.fat " VERSION " (" VERSION_DATE ")\n");
	    break;
	case 'V':
	    verify = 1;
	    break;
	case 'w':
	    write_immed = 1;
	    break;
	default:
	    usage(argv[0]);
	}
    set_dos_codepage(-1);	/* set default codepage if none was given in command line */
    if ((test || write_immed) && !rw) {
	fprintf(stderr, "-t and -w require -a or -r\n");
	exit(2);
    }
    if (optind != argc - 1)
	usage(argv[0]);

    printf("fsck.fat " VERSION " (" VERSION_DATE ")\n");
    fs_open(argv[optind], rw);

    read_boot(&fs);
    if (boot_only)
	goto exit;

    if (verify)
	printf("Starting check/repair pass.\n");
    while (read_fat(&fs), scan_root(&fs))
	qfree(&mem_queue);
    if (test)
	fix_bad(&fs);
    if (salvage_files)
	reclaim_file(&fs);
    else
	reclaim_free(&fs);
    free_clusters = update_free(&fs);
    file_unused();
    qfree(&mem_queue);
    n_files_check = n_files;
    if (verify) {
	n_files = 0;
	printf("Starting verification pass.\n");
	read_fat(&fs);
	scan_root(&fs);
	reclaim_free(&fs);
	qfree(&mem_queue);
	n_files_verify = n_files;
    }

exit:
    if (fs_changed()) {
	if (rw) {
	    if (interactive)
		rw = get_key("yn", "Perform changes ? (y/n)") == 'y';
	    else
		printf("Performing changes.\n");
	} else
	    printf("Leaving filesystem unchanged.\n");
    }

    if (!boot_only)
	printf("%s: %u files, %lu/%lu clusters\n", argv[optind],
	       n_files, fs.clusters - free_clusters, fs.clusters);

    return fs_close(rw) ? 1 : 0;
}
